### ISSUE TITLE
linux: DefaultCW: wait_pid ChildProcessError after SIGTERM

### ISSUE THREAD

**Issue by byllyfish** [byllyfish/shellous#169]
```
tests/test_shellous.py::test_asl_takewhile_sum 
-------------------------------- live log setup --------------------------------
1637257437.529 DEBUG asyncio Using selector: EpollSelector
1637257437.529 DEBUG asyncio Using selector: EpollSelector
1637257437.530 DEBUG asyncio Close <_UnixSelectorEventLoop running=False closed=False debug=True>
-------------------------------- live log call ---------------------------------
1637257437.532 INFO shellous Runner.__aenter__ stepin <Runner 'count' pid=None> (Linux-5.11.0-1021-azure-x86_64-with-glibc2.31 CPython 3.10.0 asyncio.unix_events._UnixSelectorEventLoop MainThread DefaultChildWatcher)
1637257437.532 INFO shellous Runner._start stepin <Runner 'count' pid=None>
1637257437.532 INFO shellous Runner._subprocess_spawn stepin <Runner 'count' pid=None>
1637257437.534 INFO shellous Runner._subprocess_exec stepin <Runner 'count' pid=None>
1637257437.534 DEBUG asyncio execute program '/home/runner/.cache/pypoetry/virtualenvs/shellous-dySVFLnW-py3.10/bin/python' stdin=<pipe> stdout=<pipe> stderr=<_io.TextIOWrapper name='<stderr>' mode='w' encoding='utf-8'>
1637257437.535 DEBUG shellous Thread 'EPollAgent' starting
1637257437.536 DEBUG asyncio process '/home/runner/.cache/pypoetry/virtualenvs/shellous-dySVFLnW-py3.10/bin/python' created: pid 6439
1637257437.537 DEBUG shellous _add_pidfd registered pidfd=11 pid=6439
1637257437.537 DEBUG asyncio Write pipe 12 connected: (<_UnixWritePipeTransport fd=12 idle bufsize=0>, <WriteSubprocessPipeProto fd=0 pipe=<_UnixWritePipeTransport fd=12 idle bufsize=0>>)
1637257437.539 DEBUG asyncio Read pipe 13 connected: (<_UnixReadPipeTransport fd=13 polling>, <ReadSubprocessPipeProto fd=1 pipe=<_UnixReadPipeTransport fd=13 polling>>)
1637257437.539 INFO asyncio execute program '/home/runner/.cache/pypoetry/virtualenvs/shellous-dySVFLnW-py3.10/bin/python': <_UnixSubprocessTransport pid=6439 running stdin=<_UnixWritePipeTransport fd=12 idle bufsize=0> stdout=<_UnixReadPipeTransport fd=13 polling>>
1637257437.540 INFO shellous Runner._subprocess_exec stepout <Runner 'count' pid=6439 exit_code=None> ex=None
1637257437.540 INFO shellous Runner._subprocess_spawn stepout <Runner 'count' pid=6439 exit_code=None> ex=None
1637257437.541 INFO shellous Runner._start stepout <Runner 'count' pid=6439 exit_code=None> ex=None
1637257437.541 INFO shellous Runner.__aenter__ stepout <Runner 'count' pid=6439 exit_code=None> ex=None
1637257437.541 INFO shellous read_lines stepin
1637257437.542 INFO shellous write_stream stepin
1637257437.542 INFO shellous write_stream stepout ex=None
1637257437.542 INFO shellous Runner._waiter stepin <Runner 'count' pid=6439 exit_code=None>
1637257437.569 DEBUG shellous _reap_pidfd pidfd=11 pid=6439
1637257437.569 INFO asyncio <_UnixReadPipeTransport fd=13 polling> was closed by peer
1637257437.570 INFO shellous os.waitpid(6439) returned (6439, 0)
1637257437.572 INFO shellous read_lines stepout ex=None
1637257437.573 INFO shellous Runner.__aexit__ stepin <Runner 'count' pid=6439 exit_code=None> exc_value=None
1637257437.573 INFO shellous Runner._finish stepin <Runner 'count' pid=6439 exit_code=None>
1637257437.573 INFO shellous Runner._wait stepin <Runner 'count' pid=6439 exit_code=None>
1637257437.574 INFO asyncio <_UnixSubprocessTransport pid=6439 running stdin=<_UnixWritePipeTransport closed fd=12 closed> stdout=<_UnixReadPipeTransport closed fd=13 closed>> exited with return code 0
1637257437.574 INFO shellous Runner._waiter stepout <Runner 'count' pid=6439 exit_code=0> ex=None
1637257437.575 INFO shellous Runner._wait stepout <Runner 'count' pid=6439 exit_code=0> ex=None
1637257437.575 INFO shellous Runner._close stepin <Runner 'count' pid=6439 exit_code=0>
1637257437.576 INFO shellous Runner._close stepout <Runner 'count' pid=6439 exit_code=0> ex=None
1637257437.576 INFO shellous Runner._finish stepout <Runner 'count' pid=6439 exit_code=0> ex=None
1637257437.576 INFO shellous Runner.__aexit__ stepout <Runner 'count' pid=6439 exit_code=0> ex=None exc_value=None
1637257437.576 INFO shellous Runner.__aenter__ stepin <Runner 'echo' pid=None> (Linux-5.11.0-1021-azure-x86_64-with-glibc2.31 CPython 3.10.0 asyncio.unix_events._UnixSelectorEventLoop MainThread DefaultChildWatcher)
1637257437.576 INFO shellous Runner._start stepin <Runner 'echo' pid=None>
1637257437.577 INFO shellous Runner._subprocess_spawn stepin <Runner 'echo' pid=None>
1637257437.578 INFO shellous Runner._subprocess_exec stepin <Runner 'echo' pid=None>
1637257437.578 DEBUG asyncio execute program '/home/runner/.cache/pypoetry/virtualenvs/shellous-dySVFLnW-py3.10/bin/python' stdin=<pipe> stdout=<pipe> stderr=<_io.TextIOWrapper name='<stderr>' mode='w' encoding='utf-8'>
1637257437.579 DEBUG asyncio process '/home/runner/.cache/pypoetry/virtualenvs/shellous-dySVFLnW-py3.10/bin/python' created: pid 6440
1637257437.580 DEBUG shellous _add_pidfd registered pidfd=11 pid=6440
1637257437.580 DEBUG asyncio Write pipe 12 connected: (<_UnixWritePipeTransport fd=12 idle bufsize=0>, <WriteSubprocessPipeProto fd=0 pipe=<_UnixWritePipeTransport fd=12 idle bufsize=0>>)
1637257437.581 DEBUG asyncio Read pipe 13 connected: (<_UnixReadPipeTransport fd=13 polling>, <ReadSubprocessPipeProto fd=1 pipe=<_UnixReadPipeTransport fd=13 polling>>)
1637257437.582 INFO asyncio execute program '/home/runner/.cache/pypoetry/virtualenvs/shellous-dySVFLnW-py3.10/bin/python': <_UnixSubprocessTransport pid=6440 running stdin=<_UnixWritePipeTransport fd=12 idle bufsize=0> stdout=<_UnixReadPipeTransport fd=13 polling>>
1637257437.582 INFO shellous Runner._subprocess_exec stepout <Runner 'echo' pid=6440 exit_code=None> ex=None
1637257437.582 INFO shellous Runner._subprocess_spawn stepout <Runner 'echo' pid=6440 exit_code=None> ex=None
1637257437.583 INFO shellous Runner._start stepout <Runner 'echo' pid=6440 exit_code=None> ex=None
1637257437.583 INFO shellous Runner.__aenter__ stepout <Runner 'echo' pid=6440 exit_code=None> ex=None
1637257437.584 INFO shellous read_lines stepin
1637257437.584 INFO shellous write_stream stepin
1637257437.584 INFO shellous write_stream stepout ex=None
1637257437.584 INFO shellous Runner._waiter stepin <Runner 'echo' pid=6440 exit_code=None>
1637257437.607 INFO shellous Runner.__aexit__ stepin <Runner 'echo' pid=6440 exit_code=None> exc_value=GeneratorExit()
1637257437.608 INFO shellous Runner._finish stepin <Runner 'echo' pid=6440 exit_code=None>
1637257437.608 INFO shellous Runner._kill stepin <Runner 'echo' pid=6440 exit_code=None>
1637257437.608 INFO shellous Runner.signal <Runner 'echo' pid=6440 exit_code=None> signal=<Signals.SIGTERM: 15>
1637257437.611 DEBUG shellous _reap_pidfd pidfd=11 pid=6440
1637257437.612 ERROR shellous wait_pid(6440) status is 255 ex=ChildProcessError(10, 'No child processes')
1637257437.613 INFO asyncio <_UnixSubprocessTransport pid=6440 running stdin=<_UnixWritePipeTransport closed fd=12 closed> stdout=<_UnixReadPipeTransport fd=13 polling>> exited with return code 255
1637257437.614 INFO asyncio <_UnixReadPipeTransport fd=13 polling> was closed by peer
1637257437.614 INFO shellous Runner._waiter stepout <Runner 'echo' pid=6440 exit_code=255> ex=None
1637257437.615 INFO shellous read_lines stepout ex=GeneratorExit()
1637257437.615 INFO shellous Runner._kill stepout <Runner 'echo' pid=6440 exit_code=255> ex=GeneratorExit()
1637257437.615 INFO shellous Runner._close stepin <Runner 'echo' pid=6440 exit_code=255>
1637257437.617 INFO shellous Runner._close stepout <Runner 'echo' pid=6440 exit_code=255> ex=GeneratorExit()
1637257437.617 INFO shellous Runner._finish stepout <Runner 'echo' pid=6440 exit_code=255> ex=GeneratorExit()
1637257437.617 INFO shellous Runner.__aexit__ stepout <Runner 'echo' pid=6440 exit_code=255> ex=GeneratorExit() exc_value=GeneratorExit()
PASSED
```

**Comment by byllyfish** [comment]
```
tests/test_shellous.py::test_asl_zip 
-------------------------------- live log setup --------------------------------
1637260571.614 DEBUG asyncio Using selector: EpollSelector
1637260571.615 DEBUG asyncio Using selector: EpollSelector
1637260571.615 DEBUG asyncio Close <_UnixSelectorEventLoop running=False closed=False debug=True>
-------------------------------- live log call ---------------------------------
1637260571.617 INFO shellous Runner.__aenter__ stepin <Runner 'count1' pid=None> (Linux-5.11.0-1021-azure-x86_64-with-glibc2.31 CPython 3.9.8 asyncio.unix_events._UnixSelectorEventLoop MainThread DefaultChildWatcher)
1637260571.617 INFO shellous Runner._start stepin <Runner 'count1' pid=None>
1637260571.618 INFO shellous Runner._subprocess_spawn stepin <Runner 'count1' pid=None>
1637260571.619 INFO shellous Runner._subprocess_exec stepin <Runner 'count1' pid=None>
1637260571.620 DEBUG asyncio execute program '/home/runner/.cache/pypoetry/virtualenvs/shellous-dySVFLnW-py3.9/bin/python' stdin=<pipe> stdout=<pipe> stderr=<_io.TextIOWrapper name='<stderr>' mode='w' encoding='utf-8'>
1637260571.622 DEBUG shellous Thread 'EPollAgent' starting
1637260571.631 DEBUG asyncio process '/home/runner/.cache/pypoetry/virtualenvs/shellous-dySVFLnW-py3.9/bin/python' created: pid 6334
1637260571.632 DEBUG shellous _add_pidfd registered pidfd=11 pid=6334
1637260571.633 DEBUG asyncio Write pipe 12 connected: (<_UnixWritePipeTransport fd=12 idle bufsize=0>, <WriteSubprocessPipeProto fd=0 pipe=<_UnixWritePipeTransport fd=12 idle bufsize=0>>)
1637260571.634 DEBUG asyncio Read pipe 13 connected: (<_UnixReadPipeTransport fd=13 polling>, <ReadSubprocessPipeProto fd=1 pipe=<_UnixReadPipeTransport fd=13 polling>>)
1637260571.635 INFO asyncio execute program '/home/runner/.cache/pypoetry/virtualenvs/shellous-dySVFLnW-py3.9/bin/python': <_UnixSubprocessTransport pid=6334 running stdin=<_UnixWritePipeTransport fd=12 idle bufsize=0> stdout=<_UnixReadPipeTransport fd=13 polling>>
1637260571.635 INFO shellous Runner._subprocess_exec stepout <Runner 'count1' pid=6334 exit_code=None> ex=None
1637260571.636 INFO shellous Runner._subprocess_spawn stepout <Runner 'count1' pid=6334 exit_code=None> ex=None
1637260571.637 INFO shellous Runner._start stepout <Runner 'count1' pid=6334 exit_code=None> ex=None
1637260571.637 INFO shellous Runner.__aenter__ stepout <Runner 'count1' pid=6334 exit_code=None> ex=None
1637260571.637 INFO shellous read_lines stepin
1637260571.637 INFO shellous write_stream stepin
1637260571.638 INFO shellous write_stream stepout ex=None
1637260571.638 INFO shellous Runner._waiter stepin <Runner 'count1' pid=6334 exit_code=None>
1637260571.660 INFO shellous Runner.__aenter__ stepin <Runner 'count2' pid=None> (Linux-5.11.0-1021-azure-x86_64-with-glibc2.31 CPython 3.9.8 asyncio.unix_events._UnixSelectorEventLoop MainThread DefaultChildWatcher)
1637260571.661 INFO shellous Runner._start stepin <Runner 'count2' pid=None>
1637260571.661 INFO shellous Runner._subprocess_spawn stepin <Runner 'count2' pid=None>
1637260571.662 INFO shellous Runner._subprocess_exec stepin <Runner 'count2' pid=None>
1637260571.663 DEBUG asyncio execute program '/home/runner/.cache/pypoetry/virtualenvs/shellous-dySVFLnW-py3.9/bin/python' stdin=<pipe> stdout=<pipe> stderr=<_io.TextIOWrapper name='<stderr>' mode='w' encoding='utf-8'>
1637260571.665 DEBUG shellous _reap_pidfd pidfd=11 pid=6334
1637260571.667 INFO shellous os.waitpid(6334) returned (6334, 0)
1637260571.667 DEBUG asyncio process '/home/runner/.cache/pypoetry/virtualenvs/shellous-dySVFLnW-py3.9/bin/python' created: pid 6335
1637260571.668 DEBUG shellous _add_pidfd registered pidfd=11 pid=6335
1637260571.668 INFO asyncio <_UnixSubprocessTransport pid=6334 running stdin=<_UnixWritePipeTransport closed fd=12 closed> stdout=<_UnixReadPipeTransport fd=13 polling>> exited with return code 0
1637260571.669 INFO asyncio <_UnixReadPipeTransport fd=13 polling> was closed by peer
1637260571.670 INFO shellous Runner._waiter stepout <Runner 'count1' pid=6334 exit_code=0> ex=None
1637260571.670 DEBUG asyncio Write pipe 14 connected: (<_UnixWritePipeTransport fd=14 idle bufsize=0>, <WriteSubprocessPipeProto fd=0 pipe=<_UnixWritePipeTransport fd=14 idle bufsize=0>>)
1637260571.671 DEBUG asyncio Read pipe 15 connected: (<_UnixReadPipeTransport fd=15 polling>, <ReadSubprocessPipeProto fd=1 pipe=<_UnixReadPipeTransport fd=15 polling>>)
1637260571.672 INFO asyncio execute program '/home/runner/.cache/pypoetry/virtualenvs/shellous-dySVFLnW-py3.9/bin/python': <_UnixSubprocessTransport pid=6335 running stdin=<_UnixWritePipeTransport fd=14 idle bufsize=0> stdout=<_UnixReadPipeTransport fd=15 polling>>
1637260571.672 INFO shellous Runner._subprocess_exec stepout <Runner 'count2' pid=6335 exit_code=None> ex=None
1637260571.673 INFO shellous Runner._subprocess_spawn stepout <Runner 'count2' pid=6335 exit_code=None> ex=None
1637260571.675 INFO shellous Runner._start stepout <Runner 'count2' pid=6335 exit_code=None> ex=None
1637260571.675 INFO shellous Runner.__aenter__ stepout <Runner 'count2' pid=6335 exit_code=None> ex=None
1637260571.675 INFO shellous read_lines stepin
1637260571.676 INFO shellous write_stream stepin
1637260571.676 INFO shellous write_stream stepout ex=None
1637260571.676 INFO shellous Runner._waiter stepin <Runner 'count2' pid=6335 exit_code=None>
1637260571.700 INFO shellous read_lines stepout ex=None
1637260571.700 INFO shellous Runner.__aexit__ stepin <Runner 'count1' pid=6334 exit_code=0> exc_value=None
1637260571.700 INFO shellous Runner._finish stepin <Runner 'count1' pid=6334 exit_code=0>
1637260571.700 INFO shellous Runner._wait stepin <Runner 'count1' pid=6334 exit_code=0>
1637260571.701 INFO shellous Runner._wait stepout <Runner 'count1' pid=6334 exit_code=0> ex=None
1637260571.701 INFO shellous Runner._close stepin <Runner 'count1' pid=6334 exit_code=0>
1637260571.703 INFO shellous Runner._close stepout <Runner 'count1' pid=6334 exit_code=0> ex=None
1637260571.703 INFO shellous Runner._finish stepout <Runner 'count1' pid=6334 exit_code=0> ex=None
1637260571.703 INFO shellous Runner.__aexit__ stepout <Runner 'count1' pid=6334 exit_code=0> ex=None exc_value=None
1637260571.703 INFO shellous Runner.__aexit__ stepin <Runner 'count2' pid=6335 exit_code=None> exc_value=GeneratorExit()
1637260571.704 INFO shellous Runner._finish stepin <Runner 'count2' pid=6335 exit_code=None>
1637260571.704 INFO shellous Runner._kill stepin <Runner 'count2' pid=6335 exit_code=None>
1637260571.704 INFO shellous Runner.signal <Runner 'count2' pid=6335 exit_code=None> signal=<Signals.SIGTERM: 15>
1637260571.704 DEBUG shellous _reap_pidfd pidfd=11 pid=6335
1637260571.704 ERROR shellous wait_pid(6335) status is 255 ex=ChildProcessError(10, 'No child processes')
1637260571.706 INFO asyncio <_UnixSubprocessTransport pid=6335 running stdin=<_UnixWritePipeTransport closed fd=14 closed> stdout=<_UnixReadPipeTransport fd=15 polling>> exited with return code 255
1637260571.706 INFO asyncio <_UnixReadPipeTransport fd=15 polling> was closed by peer
1637260571.707 INFO shellous Runner._waiter stepout <Runner 'count2' pid=6335 exit_code=255> ex=None
1637260571.708 INFO shellous read_lines stepout ex=GeneratorExit()
1637260571.708 INFO shellous Runner._kill stepout <Runner 'count2' pid=6335 exit_code=255> ex=GeneratorExit()
1637260571.708 INFO shellous Runner._close stepin <Runner 'count2' pid=6335 exit_code=255>
1637260571.709 INFO shellous Runner._close stepout <Runner 'count2' pid=6335 exit_code=255> ex=GeneratorExit()
1637260571.710 INFO shellous Runner._finish stepout <Runner 'count2' pid=6335 exit_code=255> ex=GeneratorExit()
1637260571.710 INFO shellous Runner.__aexit__ stepout <Runner 'count2' pid=6335 exit_code=255> ex=GeneratorExit() exc_value=GeneratorExit()
PASSED
```

**Comment by byllyfish** [comment]
This is normal behavior. Using waitid on the pidfd itself doesn't help. There is still an occasional ChildProcessErrror when we reap.

I made a fix in 0.10.0 to check for exit status 255 and re-map it to -SIGNAL if necessary.
